lib/internal/streams/destroy.js

```diff
@@ -252,6 +252,25 @@ function constructNT(stream) {
       process.nextTick(emitConstructNT, stream);
     }
   });
+  if (result !== undefined && result !== null) {
+    try {
+      const then = result.then;
+      if (typeof then === 'function') {
+        then.call(result, undefined, function(err) {
+          if (r) {
+            r.constructed = true;
+          }
+          if (w) {
+            w.constructed = true;
+          }
+          called = true;
+          process.nextTick(errorOrDestroy, stream, err);
+        });
+      }
+    } catch (err) {
+      process.nextTick(emitErrorNT, stream, err);
+    }
+  }
```

Comment:
Note that it may be worthwhile abstracting this pattern out for all of the impl functions in stream... we could even, in theory, allow an async function as an alternative to invoking the call back, e.g.

```js
class Foo extends stream.Writable {
  async _write(chunk, encoding) {
    await writeTheChunkAsynchronously();
    // returning here without an error is equivalent to invoking the callback
  }
}
```

But let's save that idea for a future pull request.